### PR TITLE
Various updates.

### PR DESCRIPTION
**Pull request by kshakir**
Switched `WdlInteger` value from java reference to scala value.
Slightly better error messages when trying to coerce `Array[Int]?` to `Array[Int?]`, or vice versa.
Cleaned up scaladoc warnings.

### CODE REVIEW

**Comment by kshakir** [comment]
FYI for reviewers-- In addition to tests in this repo, tests pass in both cromwell `sbt alltests:test` and centaur `sbt test`.

File: src/main/scala/wdl4s/values/WdlInteger.scala [hunk 1]
```diff
@@ -5,7 +5,7 @@ import wdl4s.types.WdlIntegerType
 
 import scala.util.{Failure, Success, Try}
 
-case class WdlInteger(value: Integer) extends WdlPrimitive {
+case class WdlInteger(value: Int) extends WdlPrimitive {
```
Comment (geoffjentry) [inline src/main/scala/wdl4s/values/WdlInteger.scala]: Just curious what led to this, or was it just something you noticed?

Comment (kshakir) [inline src/main/scala/wdl4s/values/WdlInteger.scala]: This is a spinoff of https://github.com/broadinstitute/cromwell/pull/2031

In this IntelliJ screenshot:
toInt implicit underlined

That implicit underlining the `.toInt` on line 178 shouldn't be there.

File: src/main/scala/wdl4s/Scope.scala [hunk 2]
```diff
@@ -181,8 +181,8 @@ trait Scope {
     * @param knownInputs All known values of FQNs
     * @param wdlFunctions Implementation of WDL functions for expression evaluation
     * @param shards For resolving specific shards of scatter blocks
-    * @return String => WdlValue lookup function rooted at `scope`
```
Comment (mcovarr) [inline src/main/scala/wdl4s/Scope.scala]: why was this rocket removed?

File: src/main/scala/wdl4s/types/WdlType.scala [hunk 3]
```diff
@@ -28,8 +29,12 @@ trait WdlType {
    */
   def coerceRawValue(any: Any): Try[WdlValue] = {
     any match {
-      case v: WdlValue if v.wdlType == this => Success(v)
-      case a if !coercion.isDefinedAt(any) => Failure(new IllegalArgumentException(s"No coercion defined from '$any' of type '${any.getClass}' to ${getClass.getSimpleName}."))
+      case wdlValue: WdlValue if wdlValue.wdlType == this => Success(wdlValue)
+      case wdlValue: WdlValue if !coercion.isDefinedAt(any) => Failure(new IllegalArgumentException(
+        s"No coercion defined from '${ScalaRunTime.stringOf(any, 3)}' of type '${wdlValue.wdlType}' to '$this'."))
```
Comment (mcovarr) [inline src/main/scala/wdl4s/types/WdlType.scala]: what is this `ScalaRunTime.stringOf` doing?  I'm not able to find docs for this.  😦

**Comment by mcovarr** [comment]
:+1:

File: src/main/scala/wdl4s/Scope.scala (hunk 2 above)
Comment (kshakir) [inline src/main/scala/wdl4s/Scope.scala]: Good catch. Overly aggressive removal. The one on the next line was causing bad scaladoc rendering [due to](https://wiki.scala-lang.org/display/SW/Syntax):
> Headings: =heading= defines a heading. Lower-level heading are obtained by using more = signs, like ===sub-heading===. A heading must be define on its own line.

producing:
```bash
[warn] /Users/kshakir/src/wdl4s/src/main/scala/wdl4s/Scope.scala:176: unbalanced or unclosed heading
[warn]   /**
[warn]   ^
```

Apparently `@return` doesn't render the same way as `@throws`, and the former rocket shows up just fine in the rendered html.

Will put it back in a squash.

File: src/main/scala/wdl4s/types/WdlType.scala (hunk 3 above)
Comment (kshakir) [inline src/main/scala/wdl4s/types/WdlType.scala]: Hehe, good catch also. `scala.runtime.ScalaRuntime` is a lot like `com.sun`, aka "this is undocumented, use this and don't complain if you get burned": https://github.com/scala/scala/blob/v2.11.8/src/library/scala/runtime/ScalaRunTime.scala#L23-L26

But it does exactly what I want, which is to properly render most `Any`s, including arrays-to-string. I didn't like the other hand-crafted suggestions in [this stackoverflow thread](http://stackoverflow.com/questions/3328085/scala-printing-arrays/3328714#3328714).

**Comment by kshakir** [comment]
Re-review please. See latest commit with changes in the error messages. Hopefully they're much simpler, but still contain enough detail for a human to debug wth is wrong.

**Comment by mcovarr** [comment]
:+1: nice, both the code and the error messages are much easier to read! 😄

File: src/main/scala/wdl4s/types/WdlType.scala [hunk 4]
```diff
@@ -28,8 +29,13 @@ trait WdlType {
    */
   def coerceRawValue(any: Any): Try[WdlValue] = {
     any match {
-      case v: WdlValue if v.wdlType == this => Success(v)
-      case a if !coercion.isDefinedAt(any) => Failure(new IllegalArgumentException(s"No coercion defined from '$any' of type '${any.getClass}' to ${getClass.getSimpleName}."))
+      case wdlValue: WdlValue if wdlValue.wdlType == this => Success(wdlValue)
+      case wdlValue: WdlValue if !coercion.isDefinedAt(any) => Failure(new IllegalArgumentException(
+        s"No coercion defined from '${wdlValue.toWdlString}' of type" +
```
Comment (cjllanwarne) [inline src/main/scala/wdl4s/types/WdlType.scala]: I personally don't like including the value here. It's almost always unreadably long and not very useful in debugging...

**Comment by cjllanwarne** [comment]
👍 otherwise

File: src/main/scala/wdl4s/types/WdlType.scala (hunk 4 above)
Comment (cjllanwarne) [inline src/main/scala/wdl4s/types/WdlType.scala]: (ToL)

**Comment by kshakir** [comment]
One more re-review please. For the coercion errors printed out, the latest commit limits the number of elements per collection to 3.

File: src/main/scala/wdl4s/values/WdlValue.scala [hunk 5]
```diff
@@ -38,7 +37,7 @@ trait WdlValue {
   def typeName: String = wdlType.getClass.getSimpleName
 
   /* This emits valid WDL source.  WdlString("foobar") -> "foobar" (quotes included) */
-  def toWdlString: String = ???
+  def toWdlString: String = throw new NotImplementedError(s"$getClass does not implement toWdlString")
```
Comment (cjllanwarne) [inline src/main/scala/wdl4s/values/WdlValue.scala]: ToL (no change req'd) - that comment on :39 seems somewhat optimistic!

**Comment by cjllanwarne** [comment]
Consider that request for change a consideration/ToL, but I do think at least something to indicate truncation (even a "ps, that might have been truncated" in the message) would be really useful to avoid confusion

**Comment by kshakir** [comment]
The ellipses identifying truncation are something I've thought about, and would be a good addition, but I don't have a quick fix for. Going to punt that for the future, as this error is appearing for an already instantiated collection, where a coercion has failed based on the expected type. The identification of the contents of the wdlValue are more hints to the user of "it was _this_ Array[Int?] that failed coercion, versus _that_ Array[Int?]".

**Comment by cjllanwarne** [comment]
re-re-👍

**Comment by mcovarr** [comment]
:+1: :+1: :+1:

**Pull request merged**